sentry-android-core/src/main/java/io/sentry/android/core/UserInteractionIntegration.java

```diff
@@ -0,0 +1,183 @@
+    final @NotNull IBuildInfoProvider buildInfoProvider,
+    final @NotNull LoadClass classLoader
+  ) {
+    this.application = Objects.requireNonNull(application, "Application is required");
+    Objects.requireNonNull(buildInfoProvider, "BuildInfoProvider is required");
+
+    isAndroidXAvailable = checkAndroidXAvailability(classLoader);
+    isAllActivityCallbacksAvailable =
+      buildInfoProvider.getSdkInfoVersion() >= Build.VERSION_CODES.Q;
+  }
+
+  private static boolean checkAndroidXAvailability(final @NotNull LoadClass loadClass) {
+    try {
+      loadClass.loadClass("androidx.core.view.GestureDetectorCompat");
+      return true;
+    } catch (ClassNotFoundException ignored) {
+      return false;
+    }
+  }
+
+  private void startTracking(final @Nullable Window window, final @NotNull Context context) {
+    if (window == null) {
+      if (options != null) {
+        options
+          .getLogger()
+          .log(SentryLevel.INFO, "Window was null in startTracking");
+      }
+      return;
+    }
+
+    if (hub != null && options != null) {
+      Window.Callback delegate = window.getCallback();
+      if (delegate == null) {
+        delegate = new NoOpWindowCallback();
+      }
+
+      final SentryGestureListener gestureListener =
+        new SentryGestureListener(new WeakReference<>(window), hub, options);
+      window.setCallback(new SentryWindowCallback(delegate, context, gestureListener, options));
+    }
+  }
+
+  private void stopTracking(final @Nullable Window window) {
+    if (window == null) {
+      if (options != null) {
+        options
+          .getLogger()
+          .log(SentryLevel.INFO, "Window was null in stopTracking");
+      }
+      return;
+    }
+
+    final Window.Callback current = window.getCallback();
+    if (current instanceof SentryWindowCallback) {
+      if (((SentryWindowCallback) current).getDelegate() instanceof NoOpWindowCallback) {
+        window.setCallback(null);
+      } else {
+        window.setCallback(((SentryWindowCallback) current).getDelegate());
+      }
+    }
+  }
+
+  @Override
+  public void onActivityPreCreated(@NotNull Activity activity,
+    @Nullable Bundle savedInstanceState) {
+    if (isAllActivityCallbacksAvailable) {
+      startTracking(activity.getWindow(), activity);
+    }
+  }
+
+  @Override
+  public void onActivityCreated(@NotNull Activity activity,
+    @Nullable Bundle bundle) {
+
+  }
+
+  @Override
+  public void onActivityStarted(@NotNull Activity activity) {
+
+  }
+
+  @Override
+  public void onActivityResumed(@NotNull Activity activity) {
+    if (!isAllActivityCallbacksAvailable) {
+      startTracking(activity.getWindow(), activity);
+    }
+  }
+
+  @Override
+  public void onActivityPaused(@NotNull Activity activity) {
+    if (!isAllActivityCallbacksAvailable) {
+      stopTracking(activity.getWindow());
+    }
+  }
+
+  @Override
+  public void onActivityStopped(@NotNull Activity activity) {
```

Comment:
Looking at this lifecycle method I assume we have another integration that already hooks into all of these.

I wonder if it wouldn't be more performatic to have a single class that inherits from `Application.ActivityLifecycleCallbacks` and that subscribe to all of these methods. And exposes a publish/subscribe hook where we can plug new integrations.

Please consider that I might be optimizing something here that doesn't require optimization. If having N instances hooked into ActivityLifecycleCallbacks is cheap, than adding the complexity clearly wouldnt' be worth it.

Reply:
Yeah, it's just a list that is being iterated whenever a callback needs to fire, so I don't think we can optimize much here

Reply:
Bruno's idea isn't bad tho, I already thought about it when doing the ActivityLifeCycle integration which does more than one thing, either separating or going with Bruno's approach is fine for me.

Reply:
filed #1877